packages/core/upload/server/controllers/content-api.js

```diff
@@ -94,8 +94,14 @@ module.exports = {
     const apiUploadFolderService = getService('api-upload-folder');
 
     const apiUploadFolder = await apiUploadFolderService.getAPIUploadFolder();
+
     data.fileInfo = data.fileInfo || {};
-    data.fileInfo = Array.isArray(data.fileInfo) ? data.fileInfo : [data.fileInfo];
+
+    data.fileInfo =
+      Array.isArray(data.fileInfo) && data.fileInfo.length === files.length
+        ? data.fileInfo
+        : Array.from({ length: files.length || 1 }, () => data.fileInfo);
```

Comment:
`files` is not always an array so the `fileInfo` may not be one neither. What do you think of this code:
```js
    const data = await validateUploadBody(body, Array.isArray(files));

    const apiUploadFolderService = getService('api-upload-folder');

    const apiUploadFolder = await apiUploadFolderService.getAPIUploadFolder();

    if (Array.isArray(files)) {
      data.fileInfo = data.fileInfo || [];
      data.fileInfo = files
        .map((f, i) => data.fileInfo[i] || {})
        .map(fileInfo => ({ ...fileInfo, folder: apiUploadFolder.id }));
    } else {
      data.fileInfo = data.fileInfo || {};
      data.fileInfo = { ...data.fileInfo, folder: apiUploadFolder.id  } ;
    }
```

Reply:
Correct me if I am wrong, but it looks like the pre-PR code was always converting `fileInfo` to an array even when there was only one file, do you know why that was happening?

Reply:
Yes you're right! The code is a bit messy around here.
Both `files` and `fileInfo` are converted to arrays just after, there: https://github.com/strapi/strapi/blob/master/packages/core/upload/server/services/upload.js#L157

I would say it's best to convert both to arrays at the same moment, so either here or later in the upload service.
I would avoid modifying the upload service to avoid the risk of breaking changes (some users may already use it).
So if we keep the array conversion in the service, I would avoid doing it here.
So technically it works but maybe it makes more sense to not convert to arrays now?

What do you think?
Or we can convert to arrays now too, it's redundant but I would not have an objection for that.

Reply:
You are right!

Just looked at the `upload` service and noticed they also get converted there.

I'll go ahead and apply your suggestion so the `upload` service remains untouched!

For tests, what would you test exactly? Maybe the length of the `fileInfo` array?

Reply:
Alright thank you!

I have in mind a e2e test : to upload multiple files and check that the files go in the correct folder
It would be located in and similar to `packages/core/upload/tests/content-api/upload-automatic-folder.test.e2e.js`

Reply:
Just pushed some tests.
I have basically re-adapted the tests from the first suite to work with multiple uploads.
Let me know what you think!